internal/resources/aws/instance.go

```diff
@@ -154,19 +154,18 @@ func (a *Instance) computeCostComponent() *schema.CostComponent {
 		}
 	}
 
+	priceFilter := &schema.PriceFilter{
+		PurchaseOption: strPtr(a.PurchaseOption),
+	}
+
+	var err error
 	if a.ReservedInstanceType != nil {
-		resolver := reservedInstanceResolver{
-			term:          strVal(a.ReservedInstanceTerm),
-			paymentOption: strVal(a.ReservedInstancePaymentOption),
-		}
-		valid, err := a.validateReserveInstanceParams(resolver)
-		if err != "" {
-			log.Warnf(err)
-		}
-		if valid {
-			purchaseOptionLabel = "reserved"
-			return a.reservedInstanceCostComponent(osLabel, osFilterVal, purchaseOptionLabel, resolver)
+		resolver := newEc2ReservationResolver(strVal(a.ReservedInstanceTerm), strVal(a.ReservedInstancePaymentOption), strVal(a.ReservedInstanceType))
+		priceFilter, err = resolver.PriceFilter()
```

Comment:
we actually only want to explicitly set the `priceFilter` if `err` is not nil. Because if it isn't then `priceFilter` will be set to nil.

Reply:
Sorry copy and paste refuse! Fixed!

Reply:
hmm this still looks the same for me, needs to be edited for all cases where `PriceFilter` is used

Reply:
Ok now I understood it better now. But now that Validation is embedded in PriceFilter how should we set priceFilter when validation fails? 
Something like this sounds good? Or we want to embed all the properties even if those fails?
```
	if r.ReservedInstanceTerm != nil {
		purchaseOptionLabel = "reserved"
		resolver := &elasticacheReservationResolver{
			term:          strVal(r.ReservedInstanceTerm),
			paymentOption: strVal(r.ReservedInstancePaymentOption),
			cacheNodeType: r.NodeType,
		}
		reservedFilter, err := resolver.PriceFilter()
		if err != nil {
			log.Warnf(err.Error())
			priceFilter = &schema.PriceFilter{
				PurchaseOption: strPtr(purchaseOptionLabel),
			}
		} else {
			priceFilter = reservedFilter
		}
	}
```

Reply:
would the "reserved" price label still be applicable if the validation failed?

Reply:
Reserved is still applicable but a warning about multiple prices found will be raised
```
"level=warning msg="Multiple prices found for aws_elasticache_cluster.redis_reserved_1yr_no_upfront ElastiCache (reserved, cache.m6g.12xlarge), using the first price"" 
```
Before of this also a validation error will be raised, as error returned from PriceFilter(), describing why validation is failing.

Reply:
Is

```go
	priceFilter := &schema.PriceFilter{
		PurchaseOption: strPtr(a.PurchaseOption),
	}
	
	if r.ReservedInstanceTerm != nil {
		purchaseOptionLabel = "reserved"
		...
		
		reservedFilter, err := resolver.PriceFilter()
		if err != nil {
			log.Warnf(err.Error())
		} else {
			priceFilter = reservedFilter
		}
	}
```

not what you want or am I missing something? Prior to this change, we did not set the `PurchaseOption` to be "reserved" so I'm unsure why you're doing that now?

Reply:
Prior to this change if `err` was not `nil` then `priceFilter` was `nil` as you mentioned. 
Adding
 ```
 ...
  } else {
	  priceFilter = reservedFilter
  }
```
If `err` is not `nil` then priceFilter will remain as set before, something like:
- https://github.com/carmeloriolo/infracost/blob/feat/elasticache-reserved-instances/internal/resources/aws/db_instance.go#L143
- https://github.com/carmeloriolo/infracost/blob/feat/elasticache-reserved-instances/internal/resources/aws/rds_cluster_instance.go#L43

So if this is not set to 'reserved' it will have an 'on_demand' value

Actually this property PurchaseOption (https://github.com/carmeloriolo/infracost/blob/feat/elasticache-reserved-instances/internal/resources/aws/instance.go#L23), is present only in instance.go

Reply:
ok yeah that was were I was getting confused, I was looking at the `db_instance` code which didn't reflect the price filter as `reserved`. Yup, what you're suggesting for `instance.go` makes sense. Alternatively, you could return a default from the `PriceFilter` method on error (for instance only) which would be something like:

```go
def :=  &schema.PriceFilter{
	PurchaseOption: strPtr("reserved"),
}

...

return def, fmt.Errorf("...")
```